### PR TITLE
[RFC] Optimize SF parameters for matecheck performance.

### PR DESCRIPTION
**Pull request by vondele**
An additional python script to optimize tunable parameters in SF for matetrack performance.

One example:

    matetrack at 1M nodes:
    Total FENs:    6554

    branch:
    Found mates:   4275
    Best mates:    3308

    master:
    Found mates:   3271
    Best mates:    2345

For a branch based on https://github.com/XInTheDark/Stockfish/commit/d83d6a4eeaa6f5efa711ade509ea3c0da8fad968

Using a budget of 4000 evaluations, invoked as:

```
mpirun -np N python3 -u -m mpi4py.futures  ng4mt.py --ngBudget 4000 --engine $dir/stockfish --epdFile $dir/matetrack.epd  --nodes 1000000 --concurrency $concurrency
```

Not sure this script belongs in the matetrack repo, the nevergrad and mpi4py dependencies are a bit heavy.

### CODE REVIEW

**Comment by robertnurnberg** [comment]
Impressive performance. What we should be careful about is to not train in the same set we test, to avoid overfitting to the test set.

So we may need a (much?) larger set of positions from which to choose the training set(s). These new positions could maybe be obtained by looking at the first mate announcement per game in pgns from fishtest? Then update the bm for those if and when we find shorter ones?

I'd be happy to keep these new scripts in the repo.

File: ng4mt.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+            print("yields: ")
+            print(output.stderr)
+            print("after: ")
+            print(output.stdout)
+            sys.exit(1)
+
+        for line in output.stdout.split("\n"):
+            # we minimize, so return minus the number of best mates
+            if "Best mates:  " in line:
+                return -int(line.split()[2])
```
Comment (robertnurnberg) [inline ng4mt.py:76]: Could also subtract a small multiple of mates found?

**Comment by vondele** [comment]
Just adding this here as a reference as well 
image

**Comment by vondele** [comment]
Note that such an optimization reduces Elo by roughly 50 points:
https://tests.stockfishchess.org/tests/view/66ac9e074ff211be9d4ed87c

**Comment by robertnurnberg** [comment]
Yeah, to be expected. But it offers some nice possibilities for the future.

File: ng4mt.py (hunk 1 above)
Comment (robertnurnberg) [reply ng4mt.py:76]: In any case I would also add `"Better mates:"`, in case the `bm` value from the epd file is not (yet) optimal. In that case could also log some message with relevant FEN.

**Comment by robertnurnberg** [comment]
A set with 1M mates has been added here: https://github.com/robertnurnberg/matetools/pull/83

Could be used to have seperate training and testing data in future.

**Comment by vondele** [comment]
this is nice functionality, but not for merging in current main branch. It can also be found back as a closed PR.
